### PR TITLE
Battles and battle reports

### PR DESCRIPTION
**Pull request by adrianlshaw**
Added basic battle support with a detailed battle report appearing in the UI.

This is basic because:
- all units have the same level of protection (called Health Points (HP)) and attack points
- there is no random seed for the encounters
- ships fire at the same time and the cleanup of ships happens after

Feel free to improve on this proof-of-concept.

### CODE REVIEW

File: app/src/main/java/dadeindustries/game/gc/mechanics/turn/Event.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+	public EventType getEventType() {
+		return type;
+	}
+
+	public String getDescription() {
+		return description;
+	}
+
+	public void appendDescription(String string) { description = description + "\n* " + string; };
```
Comment (chorhatarahuduketuri) [inline app/src/main/java/dadeindustries/game/gc/mechanics/turn/Event.java]: This should be over 3 lines, also that final semicolon on the end is unnecessary and should be removed. Autoformatting this file is a good idea.

File: app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java [hunk 2]
```diff
@@ -32,10 +35,56 @@ private void computeAiTurns(GlobalGameData globalGameData) {
 		}
 	}
 
-	public boolean processTurn(GlobalGameData globalGameData) {
+	private ArrayList<String> battleSummaries = new ArrayList<String>();
```
Comment (chorhatarahuduketuri) [inline app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java]: This field is never used.

File: app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java [hunk 3]
```diff
@@ -32,10 +35,56 @@ private void computeAiTurns(GlobalGameData globalGameData) {
...
+
+	private Event processBattle(Sector s) {
+
+		int numberOfFactions = Faction.values().length;
```
Comment (chorhatarahuduketuri) [inline app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java]: This variable is never used.

File: app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java [hunk 4]
```diff
@@ -32,10 +35,56 @@ private void computeAiTurns(GlobalGameData globalGameData) {
 		}
 	}
 
-	public boolean processTurn(GlobalGameData globalGameData) {
+	private ArrayList<String> battleSummaries = new ArrayList<String>();
+
+	private Event processBattle(Sector s) {
```
Comment (chorhatarahuduketuri) [inline app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java]: What a beautiful opportunity for some javadoc, and/or tests..... ;)

Actually unit and integration testing of the battle code is important as it's something we don't want to accidentally mess up.

File: app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java [hunk 5]
```diff
@@ -32,10 +35,56 @@ private void computeAiTurns(GlobalGameData globalGameData) {
...
+
+		int numberOfFactions = Faction.values().length;
+
+		Event result = new Event(Event.EventType.BATTLE, "Enemy ships encountered",
+				new Coordinates(s.getX(), s.getY()));
+
+		assert(s.getUnits().size() > 1);
```
Comment (chorhatarahuduketuri) [inline app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java]: Asserts are something for tests, as such this is a good example of a test case, and something that the code should handle without the program exiting. 

Additionally, it is worth considering adding the @NotNull annotation to the processBattle signature in the parameters (though I leave this up to you, as annotations can messy if you add too many).

File: app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java (hunk 4 above)
Comment (chorhatarahuduketuri) [reply app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java]: Also, s can be sector.

File: app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java [hunk 6]
```diff
@@ -32,10 +35,56 @@ private void computeAiTurns(GlobalGameData globalGameData) {
...
+			s.getUnits().get(target).damage(attack);
+
+			// Put result in an Event object and return it
+			String log = s.getUnits().get(i).getShipName() + " attacked " +
+					s.getUnits().get(target).getShipName() + " with " + attack + " damage";
+			result.appendDescription(log);
+			Log.wtf("Battle", log);
+		}
+
+		for (Unit u : s.getUnits()) {
```
Comment (chorhatarahuduketuri) [inline app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java]: u can be unit.

File: app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java [hunk 7]
```diff
@@ -50,11 +99,30 @@ public boolean processTurn(GlobalGameData globalGameData) {
 			unit.setSector(globalGameData.getSectors()[p.getX()][p.getY()]);
 			Log.wtf("Next: ", "" + p.getX() + "," + p.getY());
```
Comment (chorhatarahuduketuri) [inline app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java]: p can be pendingMove.

File: app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java (hunk 4 above)
Comment (chorhatarahuduketuri) [reply app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java]: Shouldn't this be in mechanics.units.UnitActions.java?

**Comment by chorhatarahuduketuri** [comment]
Also note that we are totally violating this: 
https://docs.oracle.com/javase/8/docs/technotes/guides/language/static-import.html

**Comment by chorhatarahuduketuri** [comment]
Actually there is something wrong with this. When I have 2 ships against 1, 2 battles happen in the same turn. Due to there being 2 battles - which are reported out of order - 4 damage is done to HMS Douglas and it is destroyed at the same time as ISS Yuri, and then after that the ISS Yuri does damage again.

**Comment by adrianlshaw** [comment]
Thanks @chorhatarahuduketuri. Fixed issue where dead ships could attack alive ships, and vice-versa. 

Haven't yet solved the issue of multiple battle reports per sector.

File: app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java (hunk 5 above)
Comment (adrianlshaw) [reply app/src/main/java/dadeindustries/game/gc/mechanics/turn/TurnProcessor.java]: Didn't know about that annotation, thanks.

**Comment by chorhatarahuduketuri** [comment]
I fixed the merge errors by reverse-merging.

**Pull request merged**